File: listing/product_type.py
"""listing/product_type.py -- every listing knows what Amazon calls it.

THE PROBLEM, IN THE OWNER'S WORDS:

    "but why are we having products with no product type, the app should be
     able to pull the product type of the items, dont skip compliance checks"

MEASURED, on his own database: 32 of 303 listings have no product_type. All 32
are on jack_uk; 30 are GENERATED, one is a variation PARENT and one is QUEUED.
Their SKUs carry eBay item ids (336475288886) where an ASIN would be, so there
was never an ASIN to read a type from, and the paths that made them -- the
variation children, the parent row, the Miles rows -- did not infer one either.

WHY A BLANK IS NOT HARMLESS. The compliance gate reads product_type to decide
whether a category can apply to this product at all. With no type it cannot
decide, and the choice is between flagging things that are not electrical and
missing things that are. Two of those 32 are "12V 10A AC to DC Adapter 120W
Power Supply" and "100W Rechargeable LED Work Light" -- both genuinely
electrical, both with an electrical flag resting on a field that is empty.

The answer is not to change how the gate treats a blank. It is to stop having
blanks.

WHERE THE ANSWER COMES FROM, BEST FIRST

  1. what the row already says          -- never overwritten
  2. Amazon, for a listing that is live -- its own answer, not ours
  3. the title                          -- infer_product_type, one shared rule
                                           set (CLAUDE.md Rule 12)

AND "HOME" IS NOT AN ANSWER. infer_product_type falls back to "HOME" so that a
submit always has a real type to send. Writing that onto a row is a different
act: the gate would then read "HOME" as a FACT about the product, and for the
12V power supply above that would turn its electrical check off -- the exact
failure this module exists to prevent. So nothing is stored unless a rule
actually matched, and a row we cannot type stays blank.
"""
import json


def from_title(title, comp_data=None):
    """The type the title implies, or "" if nothing matched.

    Deliberately thin: the rules live in amazon_listing_generator's
    _PT_INFER_RULES, which is also what the generator uses when it creates a
    listing, so a type inferred here and a type inferred there cannot differ.
    """
    try:
        from amazon_listing_generator import infer_product_type
    except Exception:
        return ""
    return infer_product_type(dict(comp_data or {}), item_name=str(title or ""),
                              default="") or ""


def from_amazon(config_path, account, marketplace, sku):
    """What Amazon itself calls this SKU. "" when it cannot say.

    Only useful for a listing that is actually on Amazon. Goes through
    api/amazon_listings, which is the one place that knows how to fetch a live
    listing -- including productTypes in includedData, without which
    product_type comes back empty for every live listing (Rule 12).
    """
    try:
        import accounts as _acc
        from api import amazon_listings as _al
    except Exception:
        return ""
    try:
        got = _al.get_item(_acc.account_creds(account), marketplace,
                           str(account.get("seller_id") or ""), sku,
                           _acc.marketplace_id(marketplace) or "")
    except Exception:
        return ""
    if not isinstance(got, dict) or got.get("status") != getattr(_al, "OK", "ok"):
        return ""
    return str(got.get("product_type") or "").strip()


def resolve(row, comp_data=None):
    """The product type for one row, without asking Amazon. "" if unknown."""
    have = str((row or {}).get("product_type") or "").strip()
    if have:
        return have
    # The attributes the generator stored are a better haystack than the title
    # alone -- item_type_keyword and the browse nodes are in there.
    cd = dict(comp_data or {})
    if not cd:
        try:
            cd = json.loads(str((row or {}).get("attributes_json") or "") or "{}")
            if not isinstance(cd, dict):
                cd = {}
            cd = {"attributes": cd}
        except Exception:
            cd = {}
    return from_title((row or {}).get("title") or "", cd)


def backfill(config_path, workspace_id):
    """Fill in the blanks for one workspace. Returns (blank_before, filled).

    NOTHING WITH A TYPE IS TOUCHED, and nothing is invented: a row whose title
    matches no rule is left blank, because a wrong type is worse than none --
    it is the difference between "we do not know" and a false statement the
    compliance gate will act on.

    Safe to run on every recompute; a row it cannot type simply stays blank and
    is tried again next time, which is what should happen after a Sync brings
    Amazon's own answer in.
    """
    from data import db as _db

    conn = _db.get_db(config_path)
    rows = [dict(r) for r in conn.execute(
        "SELECT sku, title, product_type, attributes_json FROM listings "
        "WHERE workspace_id=? AND COALESCE(product_type,'')=''", (workspace_id,))]
    if not rows:
        return 0, 0

    filled = 0
    for r in rows:
        pt = resolve(r)
        if not pt:
            continue
        conn.execute(
            "UPDATE listings SET product_type=? WHERE workspace_id=? AND sku=?",
            (pt, workspace_id, r.get("sku")))
        filled += 1
    if filled:
        conn.commit()
    return len(rows), filled


# ---- correcting drafts from Amazon's catalogue ---------------------------------
#
#     "why am i seeing that error on almost all of my listings"
#     "fix the product types from amazon"
#
# From 29 Aug 2026 every queued draft generated without its competitor ASIN, so
# its type was guessed from the title -- and HOME is what the guess says for
# almost anything (see amazon_listing_generator.process_row). Amazon then
# re-typed them at Preview: "updated from HOME to TABLE".
#
# THE ANSWER IS THE COMPETITOR ASIN'S TYPE, which is what the generator would
# have used had it been handed the ASIN. Nothing is written here: compare()
# says what WOULD change, and the owner applies it through the listing's own
# save path, so there is still exactly one thing that writes a field.

# A listing Amazon already holds is not a draft, and its type is Amazon's own
# business now. The same "on Amazon" set the barcode clash check uses (Rule 12);
# a variation PARENT and a not-yet-generated QUEUED row are not drafts either.
from domain.barcode_clash import _LIVE as _ON_AMAZON


# ---- which listings the button looks at, and why the others were left out ----
#
#     "i selected 64 listings and clicked on fix product types it shows check 2
#      unsent draft(s) against their competitor asin in amazon's catalog, - to
#      change, 2 already right 0 couldn't check"
#     "no more than 30 drafts have competitor asins in them and even if it is
#      looking at all the drafts this count is wrong and i still have that
#      warning message where amazon shows error of product type in many of the
#      items"
#
# FOUR THINGS were wrong with how the list was drawn up, and all four were silent:
#   * a draft with no competitor ASIN was dropped. There was nothing to look up,
#     which is true -- but Amazon's product-type SEARCH can say what a product
#     is from its title, and the box on the listing already uses it;
#   * a SUBMITTED listing was dropped. Right by default, never explained, and
#     impossible to override -- yet "your product type has been updated" is the
#     note Amazon attaches when it ACCEPTS a listing;
#   * the listings the owner had ticked were ignored: the whole account was read;
#   * the dialog reported only what it HAD checked, so everything left out read
#     as nothing at all.
#
# candidates() draws up the list once, and says for every listing it leaves out
# why it was left out. drafts_to_check() keeps its old meaning exactly --
# competitor-ASIN drafts only, nothing submitted -- for anything relying on it.

BY_ASIN = "asin"     # look the competitor ASIN up in Amazon's catalogue
BY_TITLE = "title"   # no ASIN: ask Amazon's product-type search about the title

SKIP_ON_AMAZON = "on_amazon"      # LIVE / ACTIVE -- its type is Amazon's now
SKIP_SUBMITTED = "submitted"      # sent to Amazon; included only on request
SKIP_PARENT = "parent"            # a variation parent is not a product
SKIP_QUEUED = "queued"            # not generated yet
SKIP_NOTHING = "nothing_to_ask"   # no competitor ASIN and no title either
SKIP_NOT_FOUND = "not_found"      # ticked, but not a draft in this account


def candidates(config_path, workspace_id, skus=None, include_submitted=False):
    """The listings Fix product types will ask Amazon about, and the rest.

    -> {"check":   [{"sku", "title", "asin", "product_type", "status",
                     "marketplace", "method"}],
        "skipped": [{"sku", "title", "status", "reason"}]}

    `skus` narrows it to the listings the owner ticked; None means the account.
    A ticked SKU this account holds no draft for is reported, not ignored.
    """
    from data import db as _db
    import re as _re
    out = {"check": [], "skipped": []}
    try:
        conn = _db.get_db(config_path)
        rows = [dict(r) for r in conn.execute(
            "SELECT sku, title, competitor_asin, product_type, status, "
            "listing_marketplace FROM listings WHERE workspace_id=?",
            (workspace_id,))]
    except Exception:
        return out

    if skus:
        wanted = [str(s).strip() for s in skus if str(s or "").strip()]
        held = {str(r.get("sku") or "").strip() for r in rows}
        for s in wanted:
            if s not in held:
                out["skipped"].append({"sku": s, "title": "", "status": "",
                                       "reason": SKIP_NOT_FOUND})
        keep = set(wanted)
        rows = [r for r in rows if str(r.get("sku") or "").strip() in keep]

    # The "on Amazon" set is barcode_clash's (Rule 12). SUBMITTED is split out of
    # it because it is the one that may be included on purpose.
    on_amazon = set(_ON_AMAZON) - {"SUBMITTED"}
    for r in rows:
        sku = str(r.get("sku") or "")
        title = str(r.get("title") or "").strip()
        status = str(r.get("status") or "").strip().upper()
        base = {"sku": sku, "title": title, "status": status}
        if status in on_amazon:
            out["skipped"].append(dict(base, reason=SKIP_ON_AMAZON))
            continue
        if status == "SUBMITTED" and not include_submitted:
            out["skipped"].append(dict(base, reason=SKIP_SUBMITTED))
            continue
        if status == "PARENT":
            out["skipped"].append(dict(base, reason=SKIP_PARENT))
            continue
        if status == "QUEUED":
            out["skipped"].append(dict(base, reason=SKIP_QUEUED))
            continue
        asin = str(r.get("competitor_asin") or "").strip().upper()
        if _re.fullmatch(r"[A-Z0-9]{10}", asin):
            method = BY_ASIN
        elif title:
            method, asin = BY_TITLE, ""
        else:
            out["skipped"].append(dict(base, reason=SKIP_NOTHING))
            continue
        out["check"].append({
            "sku": sku, "title": title, "asin": asin,
            "product_type": str(r.get("product_type") or "").strip(),
            "status": status,
            "marketplace": str(r.get("listing_marketplace") or "").strip().upper(),
            "method": method})
    return out


def drafts_to_check(config_path, workspace_id):
    """The drafts in one workspace that have a competitor ASIN to ask about.

    -> [{"sku", "title", "asin", "product_type", "status", "marketplace", "method"}]

    Unchanged in meaning: competitor-ASIN drafts only, and nothing already on
    Amazon or submitted. The wider list, with its reasons, is candidates().
    """
    return [d for d in candidates(config_path, workspace_id)["check"]
            if d["method"] == BY_ASIN]


def answer_key(draft):
    """Where compare() finds the answer for one draft.

    By ASIN for a catalogue lookup -- two drafts from one competitor share the
    answer. By SKU for a title search, because two titles are two questions.
    """
    d = draft or {}
    if d.get("method") == BY_TITLE:
        return "title:" + str(d.get("sku") or "")
    return d.get("asin")


def compare(drafts, answers):
    """What Amazon's answers would change. Pure: no database, no network.

    `answers` is {answer_key(draft): result}, where a result is
    api.amazon_catalog.product_type_of(...) for a catalogue lookup, or the same
    shape plus "options" (every type Amazon's search offered) for a title.

    -> [{"sku", "title", "asin", "current", "amazon", "verdict", "why",
         "method", "options"}]
       verdict: "change"      Amazon's type differs -- the one to apply
                "same"        already right
                "not_checked" Amazon could not answer; nothing is changed
    """
    out = []
    for d in drafts or []:
        ans = (answers or {}).get(answer_key(d)) or {}
        amazon = str(ans.get("product_type") or "").strip().upper()
        current = str(d.get("product_type") or "").strip().upper()
        if ans.get("status") != "ok" or not amazon:
            verdict, why = "not_checked", (ans.get("error") or "not asked")
        elif amazon == current:
            verdict, why = "same", ""
        else:
            verdict, why = "change", ""
        options = [str(o).strip().upper() for o in (ans.get("options") or [])
                   if str(o or "").strip()]
        out.append({"sku": d.get("sku"), "title": d.get("title"),
                    "asin": d.get("asin"), "current": current,
                    "amazon": amazon, "verdict": verdict, "why": why,
                    "method": d.get("method") or BY_ASIN, "options": options})
    return out


def still_blank(config_path, workspace_id):
    """The SKUs that even the title could not type, so a screen can say so
    rather than leaving the field quietly empty."""
    from data import db as _db
    try:
        conn = _db.get_db(config_path)
        return [r[0] for r in conn.execute(
            "SELECT sku FROM listings WHERE workspace_id=? "
            "AND COALESCE(product_type,'')=''", (workspace_id,))]
    except Exception:
        return []

File: static/js/product_type.js
/* static/js/product_type.js — Amazon's product types, from the listing screens.
 *
 *     "i see a very limited products type to select from the pdp in the app"
 *     "fix the product types from amazon, and build the search"
 *
 * TWO THINGS, both asking Amazon rather than a list the app keeps:
 *
 *   ptSearchBox   under the product-type box on a listing. Click into it and it
 *                 asks Amazon what a product with THIS TITLE is (measured: the
 *                 camping table's title returns exactly TABLE); type words and it
 *                 searches Amazon's product types for them instead.
 *
 *   ptFixDrafts   the toolbar button. Every unsent draft with a competitor ASIN
 *                 is checked against that ASIN's type in Amazon's catalogue,
 *                 and the differences are SHOWN before anything is saved.
 *
 * NOTHING HERE WRITES A FIELD ITSELF. A chosen type is saved by editField, the
 * one function that writes a listing field (Rule 12), exactly as picking from
 * the dropdown always did. The server routes are routes/product_type_routes.py.
 */

const PT_SEARCH_CACHE = {};      // "mkt|title|q" -> reply, for this page load

function _ptMkt(r){
  try{ if(typeof rowMkt === "function") return rowMkt(r); }catch(e){}
  return (typeof WS_MARKET !== "undefined" && WS_MARKET) ? String(WS_MARKET).toUpperCase() : "UK";
}

function ptSearchBox(sku, r){
  const id = "ptsx_" + sid(sku);
  return '<div class="ptsx" style="margin-top:6px">'
    + '<input id="' + id + '" class="ed" type="text" autocomplete="off" style="width:100%"'
    + ' placeholder="Search Amazon’s product types (click for suggestions from the title)"'
    + ' onfocus="ptSearchFocus(\'' + esc(sku) + '\')"'
    + ' oninput="ptSearchInput(\'' + esc(sku) + '\')">'
    + '<div id="' + id + '_out" class="cc" style="font-size:12px;margin-top:4px"></div>'
    + '</div>';
}

function _ptRow(sku){
  return (typeof ROWS !== "undefined") ? ROWS.find(x => String(x.sku) === String(sku)) : null;
}

async function _ptSearch(sku, title, q){
  const out = document.getElementById("ptsx_" + sid(sku) + "_out");
  if(!out) return;
  const r = _ptRow(sku) || {};
  const mkt = _ptMkt(r);
  const key = mkt + "|" + (title || "") + "|" + (q || "");
  out.textContent = "Asking Amazon…";
  let j = PT_SEARCH_CACHE[key];
  if(!j){
    try{
      const url = acctUrl("/product_types/search?marketplace=" + encodeURIComponent(mkt)
        + (title ? "&title=" + encodeURIComponent(title) : "")
        + (q ? "&q=" + encodeURIComponent(q) : ""));
      j = await (await fetch(url)).json();
    }catch(e){
      j = {ok:false, error:"could not reach the app"};
    }
    if(j && j.ok) PT_SEARCH_CACHE[key] = j;
  }
  // The box may have been re-typed while this was out; only paint the latest.
  const box = document.getElementById("ptsx_" + sid(sku));
  if(box && q && box.value.trim() !== q) return;
  if(!j || !j.ok){
    out.innerHTML = (j && j.status === "denied") || (j && j.denied)
      ? "Amazon does not let this account search its product types, so only the list above is available."
      : "Couldn’t search Amazon: " + esc((j && j.error) || "unknown error");
    return;
  }
  if(!(j.types || []).length){
    out.textContent = title ? "Amazon had no suggestion for this title — type some words instead."
                            : "Amazon has no product type matching those words.";
    return;
  }
  const cur = String(r.product_type || "").toUpperCase();
  out.innerHTML = (title ? "Amazon suggests for this title: " : "Amazon’s product types: ")
    + j.types.slice(0, 30).map(t =>
        '<button type="button" class="db-chip' + (t.name === cur ? " go" : "") + '"'
        + ' style="margin:2px 4px 2px 0" title="' + esc(t.display_name) + '"'
        + ' onclick="ptPick(\'' + esc(sku) + '\',\'' + esc(t.name) + '\')">'
        + esc(t.name) + (t.name === cur ? " ✓" : "") + '</button>'
      ).join("");
}

function ptSearchFocus(sku){
  const box = document.getElementById("ptsx_" + sid(sku));
  if(box && box.value.trim()) return;           // they are searching words
  const r = _ptRow(sku) || {};
  const title = String(r.title || "").trim();
  if(title) _ptSearch(sku, title, "");
}

let _PT_TIMER = null;
function ptSearchInput(sku){
  clearTimeout(_PT_TIMER);
  _PT_TIMER = setTimeout(function(){
    const box = document.getElementById("ptsx_" + sid(sku));
    const q = box ? box.value.trim() : "";
    if(q.length >= 2) _ptSearch(sku, "", q);
    else if(!q) ptSearchFocus(sku);
  }, 450);
}

/* Choosing a suggestion is choosing it in the dropdown: the same select, the
 * same onchange, so the save, the warning and the schema reload are the ones
 * that already exist. */
function ptPick(sku, name){
  const sel = document.getElementById("pt_" + sid(sku));
  if(!sel) return;
  if(![...sel.options].some(o => o.value === name)){
    const o = document.createElement("option");
    o.value = name; o.textContent = name;
    sel.insertBefore(o, sel.firstChild);
  }
  sel.value = name;
  sel.dispatchEvent(new Event("change"));
}

// ---- the toolbar button ------------------------------------------------------
//
//     "i selected 64 listings and clicked on fix product types it shows check 2
//      unsent draft(s) ... 2 already right 0 couldn't check"
//
// It read the whole account whatever was ticked, dropped every draft with no
// competitor ASIN and every submitted one, and never said so -- so 62 left out
// read as "2 already right". Now it takes the ticked listings, asks Amazon's
// product-type search about a draft with no ASIN by its title, lets you include
// submitted listings on purpose, and names every listing it leaves out and why.
// The list and the reasons are decided on the server
// (listing/product_type.candidates); this only shows them and asks.

const PT_SKIP_WHY = {
  on_amazon:      "already live on Amazon — its type is Amazon’s now",
  submitted:      "already submitted to Amazon — left out",
  parent:         "a variation parent, not a product",
  queued:         "not generated yet",
  nothing_to_ask: "no competitor ASIN and no title — nothing to ask Amazon about",
  not_found:      "ticked, but not a draft in this account",
};

/* The listings left out, grouped by reason, in the order the reasons first
 * appear. Pure, so it can be run without a page. */
function ptSkipGroups(skipped){
  const order = [], groups = {};
  (skipped || []).forEach(function(s){
    const k = (s && s.reason) || "other";
    if(!groups[k]){ groups[k] = []; order.push(k); }
    groups[k].push(s);
  });
  return order.map(function(k){ return {reason: k, why: PT_SKIP_WHY[k] || k, items: groups[k]}; });
}

/* A change is ticked for you only when there is ONE answer: the competitor
 * ASIN's type, or a title search that came back with a single type. When the
 * search offered several, you choose -- nothing is chosen on your behalf. */
function ptPreticked(r){
  if(!r || r.verdict !== "change") return false;
  if(r.method === "title") return (r.options || []).length <= 1;
  return true;
}

/* Read the results dialog: the ticked rows, each with the type chosen for it. */
function ptReadChoices(wrap, changes){
  const out = [];
  const rows = wrap ? wrap.querySelectorAll(".ptfix-row") : [];
  Array.prototype.forEach.call(rows, function(row){
    const tick = row.querySelector(".ptfix-tick");
    if(!tick || !tick.checked) return;
    const base = (changes || [])[+row.getAttribute("data-i")];
    if(!base) return;
    const pick = row.querySelector(".ptfix-pick");
    const value = pick ? String(pick.value || "").trim().toUpperCase() : String(base.amazon || "");
    if(!value) return;
    out.push(Object.assign({}, base, {amazon: value}));
  });
  return out;
}

function _ptSkipHtml(skipped){
  const groups = ptSkipGroups(skipped);
  if(!groups.length) return "";
  const n = (skipped || []).length;
  return '<details style="margin-top:8px;font-size:12px"' + (n <= 12 ? " open" : "") + '>'
    + '<summary><b>' + n + '</b> left out, and why</summary>'
    + groups.map(function(g){
        return '<div style="margin:6px 0 2px"><b>' + g.items.length + '</b> — ' + esc(g.why) + '</div>'
          + '<div class="cc" style="font-family:monospace;font-size:11px;line-height:1.5">'
          + g.items.map(function(s){ return esc(s.sku); }).join("<br>") + '</div>';
      }).join("")
    + '</details>';
}

async function _ptList(skus, includeSubmitted){
  try{
    const j = await (await fetch("/product_types/drafts", {method:"POST",
      headers:{"Content-Type":"application/json"},
      body: JSON.stringify(acctBody({skus: skus || [], include_submitted: !!includeSubmitted}))})).json();
    if(!j || !j.ok) throw new Error((j && j.error) || "refused");
    return {drafts: j.drafts || [], skipped: j.skipped || []};
  }catch(e){
    await uiAlert("Couldn’t list the drafts: " + e.message);
    return null;
  }
}

async function ptFixDrafts(){
  const aid = (typeof acctId === "function") ? acctId() : "";
  if(!aid){ toast("Open an account first"); return; }

  // THE LISTINGS YOU TICKED, when there are any -- otherwise the whole account.
  const picked = (typeof selectedSkus === "function") ? selectedSkus() : [];
  let listing = await _ptList(picked, false);
  if(!listing) return;

  // SUBMITTED LISTINGS ONLY ON PURPOSE. They are left out by default -- Amazon
  // holds them -- but "your product type has been updated" is the note Amazon
  // attaches when it accepts one, so bringing the app's type into line with
  // Amazon's is a reason to include them. Asked, never assumed.
  let includeSubmitted = false;
  const subs = listing.skipped.filter(function(s){ return s.reason === "submitted"; });
  if(subs.length){
    includeSubmitted = !!(await _dlgOpen({title:"Fix product types from Amazon",
      html:'<p style="margin:0 0 8px;font-size:12.5px"><b>' + subs.length + '</b> of '
        + (picked.length ? "the listings you ticked" : "this account’s listings") + ' '
        + (subs.length === 1 ? "is" : "are") + ' already <b>submitted</b> to Amazon.</p>'
        + '<p style="margin:0;font-size:12.5px" class="cc">They are left out unless you include them. '
        + 'Include them when Amazon has warned that it changed their product type, so this app’s '
        + 'type matches Amazon’s.</p>',
      cancelValue:false,
      buttons:[{label:"Leave them out", value:false},
               {label:"Include them", tone:"go", primary:true, value:true}]}));
    if(includeSubmitted){
      listing = await _ptList(picked, true);
      if(!listing) return;
    }
  }
  const drafts = listing.drafts, skipped = listing.skipped;
  if(!drafts.length){
    await _dlgOpen({title:"Fix product types from Amazon",
      html:'<p style="margin:0 0 8px;font-size:12.5px">Nothing to check'
        + (picked.length ? " among the " + picked.length + " listing(s) you ticked" : " in this account") + '.</p>'
        + _ptSkipHtml(skipped),
      cancelValue:null, buttons:[{label:"Close", tone:"go", primary:true, value:null}]});
    return;
  }

  let cancelled = false;
  _dlgOpen({title:"Fix product types from Amazon",
            html:'<div id="ptfix_prog" style="font-size:12.5px">Asking Amazon…</div>',
            cancelValue:null, buttons:[{label:"Cancel", value:null}]})
    .then(function(){ cancelled = true; });
  const prog = t => { const el = document.getElementById("ptfix_prog"); if(el) el.textContent = t; };

  // A few drafts per request, per marketplace. The verdicts come back decided
  // -- listing/product_type.compare() is the one definition of "change".
  const byMkt = {};
  drafts.forEach(d => {
    const m = d.marketplace || _ptMkt({});
    (byMkt[m] = byMkt[m] || []).push(d);
  });
  const rows = [];
  let done = 0, denied = false;
  for(const mkt of Object.keys(byMkt)){
    const list = byMkt[mkt];
    for(let i = 0; i < list.length; i += 15){
      if(cancelled) return;
      prog("Asking Amazon about " + drafts.length + " draft(s)… " + done + " done");
      const batch = list.slice(i, i + 15);
      let j;
      try{
        j = await (await fetch("/product_types/lookup", {method:"POST",
          headers:{"Content-Type":"application/json"},
          body: JSON.stringify(acctBody({marketplace: mkt, skus: batch.map(d => d.sku),
                                         include_submitted: includeSubmitted}))})).json();
      }catch(e){ j = {ok:false, error:"could not reach the app"}; }
      if(j.ok){
        (j.rows || []).forEach(r => rows.push(r));
        if(j.denied) denied = true;
      }else{
        if(j.denied) denied = true;
        batch.forEach(d => rows.push({sku:d.sku, title:d.title, asin:d.asin, method:d.method,
          current:d.product_type, amazon:"", verdict:"not_checked", why:j.error || "failed", options:[]}));
      }
      done += batch.length;
    }
  }
  if(cancelled) return;
  const changes = rows.filter(r => r.verdict === "change");
  const same = rows.filter(r => r.verdict === "same").length;
  const unchecked = rows.filter(r => r.verdict === "not_checked");
  const byAsin = rows.filter(r => r.method !== "title").length;
  const byTitle = rows.length - byAsin;

  let html = '<p style="margin:0 0 8px;font-size:12.5px">'
    + (picked.length ? "You ticked <b>" + picked.length + "</b> listing(s). " : "")
    + 'Checked <b>' + rows.length + '</b>: ' + byAsin + ' by competitor ASIN, '
    + byTitle + ' by title (Amazon’s product-type search).</p>'
    + '<p style="margin:0 0 8px;font-size:12.5px"><b>' + changes.length + '</b> to change'
    + ' · ' + same + ' already right · ' + unchecked.length + ' couldn’t check'
    + ' · ' + skipped.length + ' left out</p>';
  if(denied){
    html += '<p style="margin:0 0 8px;font-size:12px" class="cwarn">Amazon refused this account’s access for some of these, so they were not checked and will not be changed.</p>';
  }
  if(changes.length){
    html += '<div style="max-height:40vh;overflow-y:auto;font-size:12px;border-top:1px solid var(--line2)">'
      + changes.map(function(r, i){
          const opts = r.options || [];
          const choose = (r.method === "title" && opts.length > 1)
            ? '<select class="ptfix-pick" style="font-size:12px" onchange="this.closest(\'.ptfix-row\').querySelector(\'.ptfix-tick\').checked=true">'
              + opts.map(o => '<option value="' + esc(o) + '"' + (o === r.amazon ? " selected" : "") + '>' + esc(o) + '</option>').join("")
              + '</select>'
            : '<b>' + esc(r.amazon) + '</b>';
          return '<label class="ptfix-row" data-i="' + i + '" style="display:flex;gap:8px;padding:5px 0;border-bottom:1px solid var(--line2)">'
            + '<input type="checkbox" class="ptfix-tick"' + (ptPreticked(r) ? " checked" : "") + '>'
            + '<span style="flex:1"><span style="font-family:monospace">' + esc(r.sku) + '</span>'
            + ' <span class="cc">' + (r.method === "title" ? "· suggested from title" : "· from ASIN " + esc(r.asin || "")) + '</span>'
            + '<div class="cc">' + esc(String(r.title || "").slice(0, 70)) + '</div>'
            + '<div>' + esc(r.current || "(blank)") + ' → ' + choose + '</div></span></label>';
        }).join("") + '</div>'
      + '<p style="margin:8px 0 0;font-size:12px" class="cc">Title suggestions are Amazon’s best guess — check them. '
      + 'Where Amazon offered several types, pick one and tick the row. Nothing is saved until you press Apply. '
      + 'A different type can ask for different details: run Preview on these afterwards.</p>';
  }
  if(unchecked.length){
    html += '<details style="margin-top:8px;font-size:12px"><summary>Couldn’t check (' + unchecked.length + ')</summary>'
      + unchecked.map(r => '<div>' + esc(r.sku) + ' — ' + esc(r.why) + '</div>').join("") + '</details>';
  }
  html += _ptSkipHtml(skipped);
  const chosen = await _dlgOpen({title:"Fix product types from Amazon", html:html, cancelValue:null,
    buttons: changes.length
      ? [{label:"Cancel", value:null},
         {label:"Apply ticked changes", tone:"go", primary:true,
          take: function(wrap){
            const got = ptReadChoices(wrap, changes);
            if(!got.length){ toast("Tick at least one change"); return undefined; }
            return got;
          }}]
      : [{label:"Close", tone:"go", primary:true, value:null}]});
  if(!chosen || !chosen.length) return;

  let ok = 0; const failed = [];
  for(const r of chosen){
    const j = await editField(r.sku, "col", "Product Type", r.amazon);
    if(j && j.ok) ok++; else failed.push(r.sku + ": " + ((j && j.error) || "save failed"));
  }
  try{
    await fetch("/product_types/recheck", {method:"POST", headers:{"Content-Type":"application/json"},
      body: JSON.stringify(acctBody({}))});
  }catch(e){ /* the saves stand; warnings catch up on the next generate */ }
  if(typeof refreshView === "function"){ try{ refreshView(); }catch(e){} }
  // The warnings on each listing's page are Amazon's reply to its LAST Preview or
  // Submit, and only a new one replaces them -- say so, or a saved fix looks
  // like a failed one ("it says it is fixed 33 but ... that error message still").
  await uiAlert("Changed " + ok + " product type(s)."
    + (failed.length ? "\n\nNot saved:\n" + failed.join("\n") : "")
    + (ok ? "\n\nAmazon's old warnings stay on these listings until they are previewed again. "
          + "Tick them and press Preview on the selection bar to refresh them all at once." : ""));
}
